File: python/sql.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

"""
1 python DBI 数据库连接对象的获得一个游标对象；
2 游标的execute()向服务器发送语句，抛出异常或返回结果集
"""


## 	A 不返回结果
##	INSERT
##	DELETE
##	UPDATE
cursor = conn.cursor()
cursor.execute("UPDATE profile SET cats = cats + 1 WHERE name = 'Fred'")
print "Number of rows updated: %d" % cursor.rowcount
cursor.close()


## 	B 返回结果
##	SELECT
##	SHOW
##	EXPLAIN
##	DESCRIBE
##	fetchone() 顺序返回下一行
cursor = conn.cursor()
cursor.execute("SELECT id, name, cats FROM profile")
while 1 :
	row = cursor.fetchone() #len(row)
	if row == None :
		break
	print "id: %s, name: %s, cats:%s" % (row[0],row[1],row[2])
print "Number of rows returned: %d" % cursor.rowcount
cursor.close()

##	fetchall() 行序列方式返回整个结果集
cursor = conn.cursor()
cursor.execute("SELECT id, name, cats FROM profile")
rows= cursor.fetchall() #len(rows) 
for row in rows:  #rows[1][2]
	print "id: %s, name: %s, cats:%s" % (row[0],row[1],row[2])
print "Number of rows returned: %d" % cursor.rowcount
cursor.close()


##	DictCursor通过列名来访问行值
cursor = conn.cursor(MySQLdb.cursor.DictCursor)
cursor.execute("SELECT id, name, cats FROM profile")
for row in cursor.fetchall():
	print "id: %s, name: %s, cats:%s" % (row[0],row[1],row[2])
print "Number of rows returned: %d" % cursor.rowcount
cursor.close()

File: perl/sql.pl
## 	A 不返回结果
##	INSERT
##	DELETE
##	UPDATE
my $count = $dbh->("UPDATE profile SET cats = cats + 1 WHERE name = 'Fred' ");
if ($count) { # 如果发生错误则输出行数
    $count += 0; # 0E0
    print "Number of row updated: %d\n", $count;
}

## 	B 返回结果
##	SELECT
##	SHOW
##	EXPLAIN
##	DESCRIBE
##	1.首先,prepare()执行语句,
##		成功:	返回用于后续操作的statement句柄，
##		失败:	停止或返回undef
##	2.调用execute()执行语句，生成结果集
##	3.循环语句获取返回的行
##	4.finish() 释放资源

#1
my $sth = $dbh->prepare("SELECT id, name, cats FROM profile"); 
#2
$sth->execute();
my $count;
#3
# fetchrow_array()显示的定义select语句
# my $ref = $sth->fetchrow_arrayref()  通过$ref->[0]来访问
# my $ref = $sth->fetchrow_hashref()   通过$ref->{id}来访问，列名需唯一
while (my @val = $sth->fetchrow_array()) {
	print "id:$val[0], name:$val[1],cats:$val[2]\n";
	++$count;
}
#4
$sth->finish();
print "Number of rows returned: %d\n", $count;


##	数据库句柄方法
#	selectrow_array() 
#	selectrow_arrayref()
#	selectcol_arrayref()
#	selectall_arrayref()
#	selectrow_hashref()
#	selectall_hashref()
my @val = $dbh->selectrow_array("SELECT id, name, cats FROM profile WHERE id = 3"); 

